### DISCUSSION TITLE
Shell, logout, and login again don't create a new viewmodel

### DISCUSSION THREAD

**Discussion by softlion** [xamarin/Xamarin.Forms#15174] [Q&A]
Hi mates,
i have a simple scenario involving a login and a tabbed home screen with a logout command.

Issues:
1. after log in, opening the home tabbar does not remove the login page from the navigation hierarchy. I used RemovePage to remove the pages in the navigationstack as a workaround. Is that mandatory ?
2. after logging out and log in again with another user, the home still displays the informations of the logged out user, because the home view/viewmodel are not recreated but reused instead. How to force the re-creation of the view/viewmodel ? Why is it still in memory ?

AppShell.xml code:
```xml
<ShellContent Route="LoginPage" ContentTemplate="{DataTemplate local:LoginPage}"/>

<TabBar Route="home">
    <ShellContent Route="HomePage" ContentTemplate="{DataTemplate home:HomePage}" Title="Home" />
    <ShellContent Route="OtherPage" ContentTemplate="{DataTemplate home:OtherPage}" Title="Other" />
</TabBar>
```

1. code to login
```c#
//Don't know how to clear the stack after navigating away
foreach (var page in Shell.Current.Navigation.NavigationStack.Reverse().Where(p => p != null))
    Shell.Current.Navigation.RemovePage(page);
await Shell.Current.GoToAsync($"//home/{nameof(HomePage)}");
```

2. code to logout
```c#
await Shell.Current.GoToAsync($"//LoginPage");
```

**Comment by MaxFelinger** [comment]
Having the same problem. 
I was going even one step deeper and also removed everything from the NavigationStack of all Tabs. But the ViewModel was still the same instance as before logging out.
PS: I am also using DataTemplates.

The only workaround I can see is to remove all unnecessary Tabs on Logout.

**Reply by softlion** [reply]
Looking at the Shell's tab source code, you can see that the tabs viewmodels are cached. And this cache is static. And never cleared.
The official answer is "it is by design"...
I don't know who is the designer, but he should not build mobile apps often ...

**Reply by MaxFelinger** [reply]
True, it's unfortunate that the cache was made static.

But back to the search for a possible workaround. I can see that ShellContent is cached, which is a sub item of a TabItem. The TabItem itself is a ShellSection which is not cached. Removing and adding TabItems works for me. The TabItemRootPageViewModel is a new instance after removing and adding again.

**Reply by softlion** [reply]
Ok cool a workaround. I prefer not to have workaround in a product which was released less than 10 days ago, and for which I notify about this issue in march 2022, but also in december 2021.
